probe_src/probe_frontend/src/display.rs

```diff
@@ -0,0 +1,302 @@
+use std::fmt::Display;
+
+use crate::ops;
+use chrono::{DateTime, SecondsFormat};
+
+impl Display for ops::statx_timestamp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match DateTime::from_timestamp(self.tv_sec, self.tv_nsec) {
+            Some(x) => f.write_str(&x.to_rfc3339_opts(SecondsFormat::Secs, true)),
+            None => f.write_str("[INVALID TIMESTAMP]"),
+        }
+    }
+}
+
+impl Display for ops::timeval {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match DateTime::from_timestamp(self.tv_sec, self.tv_usec as u32 * 1000) {
+            Some(x) => f.write_str(&x.to_rfc3339_opts(SecondsFormat::Secs, true)),
+            None => f.write_str("[INVALID TIMESTAMP]"),
+        }
+    }
+}
+
+impl Display for ops::statx {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f,
+            "[ mask={}, blksize={}, attributes={}, nlink={}, uid={}, gid={}, \
+            mode={:#06o} ino={}, size={}, blocks={}, attributes_mask={}, \
+            atime={}, btime={}, ctime={}, mtime={}, rdev_major={}, \
+            rdev_minor={}, dev_major={}, dev_minor={}, mnt_id={}, \
+            dio_mem_align={}, dio_offset_align={} ]",
+            self.stx_mask,
+            self.stx_blksize,
+            self.stx_attributes,
+            self.stx_nlink,
+            self.stx_uid,
+            self.stx_gid,
+            self.stx_mode,
+            self.stx_ino,
+            self.stx_size,
+            self.stx_blocks,
+            self.stx_attributes_mask,
+            self.stx_atime,
+            self.stx_btime,
+            self.stx_ctime,
+            self.stx_mtime,
+            self.stx_rdev_major,
+            self.stx_rdev_minor,
+            self.stx_dev_major,
+            self.stx_dev_minor,
+            self.stx_mnt_id,
+            self.stx_dio_mem_align,
+            self.stx_dio_offset_align,
+        )
+    }
+}
+
+impl Display for ops::rusage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f,
+            "[ utime={}, stime={}, maxrss={}, ixrss={}, idrss={}, isrss={}, \
+            minflt={}, majflt={}, nswap={}, inblock={}, oublock={}, msgsnd={}, \
+            msgrcv={}, nsignals={}, nvcsw={}, nivcsw={} ]",
+            self.ru_utime,
+            self.ru_stime,
+            self.ru_maxrss,
+            self.ru_ixrss,
+            self.ru_idrss,
+            self.ru_isrss,
+            self.ru_minflt,
+            self.ru_majflt,
+            self.ru_nswap,
+            self.ru_inblock,
+            self.ru_oublock,
+            self.ru_msgsnd,
+            self.ru_msgrcv,
+            self.ru_nsignals,
+            self.ru_nvcsw,
+            self.ru_nivcsw,
+        )
+    }
+}
+
+impl Display for ops::Path {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f,
+            "[ dirfd_minus_at_fdcwd={}, path='{}', device_major={}, \
```

Comment:
We should definitely hide `dirfd_minus_at_fdcwd` and only print `dirfd_minus_at_fdcwd + AT_FDCWD`. This is a hack to make `dirfd` storable in an unsigned type despite the fact that `AT_FDCWD` is usually defined as -100 in Linux. We can't have users knowing how hacky the code is ;P

We should consider just printing `path`, `dirfd`, `inode`.

Reply:
lmao, I won't pretend I checked what every field did for every Op, but that makes sense, will do.

Reply:
Addressed by https://github.com/charmoniumQ/PROBE/pull/20/commits/6573d3cec29bdfb5f4c9027b142ba260ec59bb6e.